tee-worker/service/src/main_impl.rs

```diff
@@ -521,86 +521,51 @@ fn start_worker<E, T, D, InitializationHandler, WorkerModeProvider>(
-						if value.mr_enclave.to_vec() == mrenclave && value.url == trusted_url {
-							// After calling the perform_ra function, the nonce will be incremented by 1,
-							// so enclave is already registered, we should reset the nonce_cache
-							let nonce =
-								litentry_rpc_api.get_account_next_index(&tee_accountid).unwrap();
-							enclave
-								.set_nonce(nonce, ParentchainId::Litentry)
-								.expect("Could not set nonce of enclave. Returning here...");
-							found = true;
-							info!("fond enclave: {:?}", value);
-							break
-						}
-					},
-					Ok(None) => {
-						warn!("not found from key: {:?}", key);
-					},
-					Err(_) => {},
-				}
-			}
-			if !found {
-				// Todo: Can't unwrap here because the extrinsic is for some reason not found in the block
-				// even if it was successful: https://github.com/scs/substrate-api-client/issues/624.
-				let register_enclave_block_hash = send_register_xt();
-				let api_register_enclave_xt_header =
-					litentry_rpc_api.get_header(register_enclave_block_hash).unwrap().unwrap();
-
-				// TODO: #1451: Fix api-client type hacks
-				// TODO(Litentry): keep an eye on it - it's a hacky way to convert `SubstrateHeader` to `Header`
-				let header =
-					Header::decode(&mut api_register_enclave_xt_header.encode().as_slice())
-						.expect("Can decode previously encoded header; qed");
-
-				println!(
-					"[+] Enclave registered at block number: {:?}, hash: {:?}",
-					header.number(),
-					header.hash()
-				);
+	// Litentry: send the registration extrinsic regardless of being registered or not,
+	//           the reason is the mrenclave could change in between, so we rely on the
+	//           on-chain logic to handle everything.
+	//           this is the same behavior as upstream
+	let register_enclave_block_hash =
+		send_register_xt().expect("enclave RA registration must be successful to continue");
+
+	let api_register_enclave_xt_header =
+		litentry_rpc_api.get_header(Some(register_enclave_block_hash)).unwrap().unwrap();
+
+	// TODO: #1451: Fix api-client type hacks
+	let register_enclave_xt_header =
+		Header::decode(&mut api_register_enclave_xt_header.encode().as_slice())
+			.expect("Can decode previously encoded header; qed");
+
+	println!(
+		"[+] Enclave registered at block number: {:?}, hash: {:?}",
+		register_enclave_xt_header.number(),
+		register_enclave_xt_header.hash()
+	);
 
-				register_enclave_xt_header = Some(header);
-			}
-		},
-		_ => panic!("unknown error"),
-	}
+	// double-check
+	let my_enclave = litentry_rpc_api
+		.enclave(&tee_accountid, None)
+		.unwrap()
+		.expect("our enclave should be registered at this point");
+	trace!("verified that our enclave is registered: {:?}", my_enclave);
 
-	if let Some(register_enclave_xt_header) = register_enclave_xt_header.clone() {
-		we_are_primary_validateer =
-			we_are_primary_worker(&litentry_rpc_api, &register_enclave_xt_header).unwrap();
-	}
+	let is_primary_enclave = match litentry_rpc_api
```

Comment:
It's just a little optimization but maybe the `Teebag` `EnclaveAdded` event could contain information whether it was the primary enclave registered or not.

Then we could look up for this event in [`ExtrinsicReport`](https://github.com/scs/substrate-api-client/blob/master/src/api/mod.rs#L39) and this additional api call(s) could be omitted.

Reply:
I think that's actually more complicated 😂 You need to verify it's the first `EnclaveAdded` event you've got?